### PR TITLE
Display subworkflows (fc-app#11)

### PR DESCRIPTION
**Pull request by jmthibault79**
- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: If you changed a URL that is used elsewhere (e.g. in an email), comment about where it is used and ensure the dependent code is updated.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [ ] **Submitter**: If you're adding new automated UI tests, review the test plan with QA
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Submitter**: Verify all tests go green, including CI tests and automated UI tests.
- [ ] **Submitter**: Squash commits and merge to develop. If adding test code, merge application code and test code at the same time.
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

**Comment by jmthibault79** [comment]
Half of this is whitespace so try this on for size: https://github.com/broadinstitute/firecloud-ui/pull/1295/files?w=1

File: src/cljs/main/broadfcui/page/workspace/monitor/workflow_details.cljs [hunk 1]
```diff
@@ -177,94 +177,112 @@
...
-         (:data props)))])})
+     (let [subWorkflowId (:subWorkflowId props)]
+       [:span {}
+        (links/create-internal {:onClick #(swap! state update :expanded not)}
+                               (if (:expanded @state) "Hide" "Show"))
+        [:div {}
+         (when (:expanded @state)
+           [WorkflowDetails (merge props {:workflow-id subWorkflowId})])]]))})
 
+(defn- render-gcs-path [components]
```
Comment (jmthibault79) [inline src/cljs/main/broadfcui/page/workspace/monitor/workflow_details.cljs]: Example Input: 
```
[fc-13226cd9-8f18-4969-a7fd-95dc367ee5e1,
33c68fba-3cef-40a2-9e5e-8ecffe6d8750,
scatterHC_samples,
554acfd2-769e-4a60-b5d2-3d79bb3cb904]
```

Output: https://accounts.google.com/AccountChooser?continue=https://console.cloud.google.com/storage/browser/fc-13226cd9-8f18-4969-a7fd-95dc367ee5e1/33c68fba-3cef-40a2-9e5e-8ecffe6d8750/scatterHC_samples/554acfd2-769e-4a60-b5d2-3d79bb3cb904/

Comment (jmthibault79) [reply src/cljs/main/broadfcui/page/workspace/monitor/workflow_details.cljs]: This allows me to recursively build the GCS paths.  After https://github.com/broadinstitute/cromwell/issues/3623 we can discard this and use the `workflowRoot` value in the metadata directly.

File: src/cljs/main/broadfcui/page/workspace/monitor/workflow_details.cljs [hunk 2]
```diff
@@ -177,94 +177,112 @@
...
+                (create-field "stderr" (display-value (data "stderr") (last (string/split (data "stderr") #"/"))))
+                (backend-logs data)
+                (when-let [failures (data "failures")]
+                  [Failures {:data failures}])]])
+            (:data props)))]))})
 
+(defn- render-workflow-detail [workflow raw-data workflow-name submission-id bucketName workspace-id gcs-path-prefix]
+  (let [inputs (ffirst (workflow "calls"))
+        input-names (string/split inputs ".")
+        workflow-name-for-path (first input-names)
```
Comment (jmthibault79) [inline src/cljs/main/broadfcui/page/workspace/monitor/workflow_details.cljs:303]: This was conflicting with the other `workflow-name` when I expanded its scope

**Comment by jmthibault79** [comment]
Partially depends on and partially copies #1297 so let's merge that one first.

**Review by davidangb — APPROVED** [review]
:+1: looks good, works good as far as code goes. I have some functional questions that I will ask outside of the PR.

File: src/cljs/main/broadfcui/page/workspace/monitor/workflow_details.cljs [hunk 3]
```diff
@@ -177,92 +177,156 @@
...
+       [WorkflowTiming {:label "Workflow Timing" :data raw-data :workflow-name workflow-name}])
+     (when-let [failures (workflow "failures")]
+       [Failures {:data failures}])
+     (when (seq (workflow "calls"))
+       [:div {:style {:marginTop "1em" :fontWeight 500}} "Calls:"])
+     (for [[call data] (workflow "calls")]
+       [CallDetail {:label call :data data :submission-id submission-id :workflowId (workflow "id")
+                    :workspace-id workspace-id :gcs-path-prefix subworkflow-path-components}])]))
+
+(react/defc- SubworkflowDetails
```
Comment (jmthibault79) [inline src/cljs/main/broadfcui/page/workspace/monitor/workflow_details.cljs:208]: I'd really like to reuse `WorkflowDetails` but haven't been able to get that to work.  The only differences are choosing which `render-*-detail` function to use and setting the `gcs-path-prefix`

**Review by davidangb — APPROVED** [review]
:+1: approved!

For the sake of discussion and not necessary at all for this PR (in fact I advise against it, ROI may not be worth it): every time I hide/show a subworkflow, the app makes a new ajax call to the server. I assume this is because the `(when ... )` statement is mounting/unmounting the component, and therefore the `:component-did-mount` is firing. What would be preferable is:
- if component doesn't exist, mount it and load its data
- once it exists, just toggle `display: none` vs. `display: block` styles so it hides/shows without going through React lifecycle

I don't know the best way to do this, and as I said it feels like an optimization that isn't worth the effort, at least right now.

**Comment by jmthibault79** [comment]
True, though I wonder if repeat Show/Hides would be a common user behavior?

From a pragmatic standpoint, I'm not familiar with how `display` works with React so there would be a learning curve.

**Pull request merged**